### PR TITLE
Use `Expr` enum instead of `Box<dyn Expression>`

### PR DESCRIPTION
**Pull request by Sladuca**
`Box<dyn Expression>` is annoying to use during tests. Enums are nice because you can match against them. So (as discussed OOB) I created an enum type `Expr` that's an enum over all possible kinds of expressions.

I also cleaned up the module hierarchy and derived some traits for `Expr` that makes it easier to write parser tests.

### CODE REVIEW

**Comment by encody** [comment]
I personally prefer to avoid `mod.rs` files because as the number of modules increases it becomes difficult to know _which_ `mod.rs` I'm looking at. Thoughts?

File: ast/src/expression/bind.rs [hunk 1]
```diff
@@ -115,14 +117,14 @@ mod tests {
...
-            body: Box::new(Apply(
-                Box::new(Apply(
-                    Box::new(Binary::Equals),
-                    Box::new(Primitive::UInt64(5)),
-                )),
-                Box::new(Rvalue(Var::Bind("x".to_string()))),
-            )),
+            value: Box::new(Expr::Primitive(Primitive::UInt64(5))),
+            body: Box::new(Expr::Apply(Apply(
+                Box::new(Expr::Apply(Apply(
```
Comment (encody) [inline ast/src/expression/bind.rs:122]: although i agree with the design decisions necessitating this verbosity, i hate it. imma write some expression generator macros

**Review by encody — APPROVED** [review]

**Comment by encody** [comment]
Actually, this gives us an opportunity to reduce the absolute number of `Box`es used, give me a minute...

**Comment by encody** [comment]
https://github.com/encody/rusteal/pull/6

File: ast/src/expression/mod.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            Expr::OnComplete(expr) => expr.compile(context, prepared_stack),
+            Expr::Primitive(expr) => expr.compile(context, prepared_stack),
+            Expr::Ret(expr) => expr.compile(context, prepared_stack),
+            Expr::Seq(expr) => expr.compile(context, prepared_stack),
+            Expr::Txn(expr) => expr.compile(context, prepared_stack),
+            Expr::LVal(expr) => expr.compile(context, prepared_stack),
+            Expr::RVal(expr) => expr.compile(context, prepared_stack),
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (Sladuca) [reply ast/src/expression/mod.rs:82]: I could use a macro

**Comment by Sladuca** [comment]
> I personally prefer to avoid mod.rs files because as the number of modules increases it becomes difficult to know which mod.rs I'm looking at. Thoughts?

It's based on the containing folder. This is the standard way rust crates are written and it's easier to follow for people used to working on / with other rust crates.

**Pull request merged**